### PR TITLE
feat(3666): refactor quoting service into an event driven solution to improve performance

### PR DESCRIPTION
**Pull request by geka-evk**
feat(3666): refactor quoting service into an event driven solution to improve performance.

_The current quote service performs this action in memory, and a better solution would be to split the servce into two microservices with Kafka events. This will decouple the participant API responses from the work being performed and free's up the nodeJS IO threads._
- added Kafka config
- updated /quotes  /bulkQuotes routes
- added QuotingHandler
- updated tests

### CODE REVIEW

File: src/handlers/init.js
Comment (vijayg10) [inline src/handlers/init.js:33]: There is file setup.js which handles the registration of handlers in central ledger. It also abstracts some other functionalities like running migration scripts, initializing cache ...etc. Can we follow the same approach here instead of calling starting consumer here directly?

File: config/default.json
Comment (geka-evk) [reply config/default.json:78]: https://github.com/mojaloop/central-ledger/blob/main/config/default.json#L203

I see we use singular for _transfer_, so I followed the same approach

File: config/default.json
Comment (geka-evk) [reply config/default.json:155]: https://github.com/mojaloop/central-ledger/blob/main/config/default.json#L203

I see we use singular for _transfer_, so I followed the same approach

File: config/default.json
Comment (geka-evk) [reply config/default.json:234]: https://github.com/mojaloop/central-ledger/blob/main/config/default.json#L203

I see we use singular for _transfer_, so I followed the same approach

File: config/default.json
Comment (geka-evk) [reply config/default.json:296]: https://github.com/mojaloop/central-ledger/blob/main/config/default.json#L203

I see we use singular for _transfer_, so I followed the same approach

File: src/api/bulkQuotes/{id}/error.js
Comment (geka-evk) [reply src/api/bulkQuotes/{id}/error.js:88]: coz we moved actual quotes/bulkQuotes processing to consumers. Si auditing is used there:
 - https://github.com/mojaloop/quoting-service/pull/321/files#diff-d24997468f7bff06c3e451183836b6c5131fd856b5ff26e8e8eb2412759997a7R186
 - https://github.com/mojaloop/quoting-service/pull/321/files#diff-d24997468f7bff06c3e451183836b6c5131fd856b5ff26e8e8eb2412759997a7R125

File: src/handlers/QuotingHandler.js
Comment (geka-evk) [reply src/handlers/QuotingHandler.js:55]: As far as we agreed to publish POST/PUT/GET quotes/bulkQuotes payloads to a separate queues, what's wrong with having a separate processing function per queue?  Do you think, in topic `topic-quotes-post` we might have a message with type `bulkQueue` and action `get`?

File: src/handlers/QuotingHandler.js
Comment (geka-evk) [reply src/handlers/QuotingHandler.js:83]: Sorry, but I don't get what benefits having a separate handlePutQuotesError function gives us. We publish PUT suceess and error payload in the same queue, so what's wrong with handling of 2 types of payload in one processing functions. Also, please take into account, that actual processing is done in a separate model methods

File: jest.config.js
Comment (geka-evk) [reply jest.config.js:11]: I focused on adding unit-tests, first of all, for business logic. So that I could start testing my changes with real Kafka and DB earlier (to be sure that I didn't introduce any regression). But if we have enough time to cover all these "helper" functionalities, let's do it, I don't mind.

File: src/handlers/init.js
Comment (geka-evk) [reply src/handlers/init.js:33]: First of all, there is no step for running migration during handlers start:
https://github.com/mojaloop/central-ledger/blob/main/src/handlers/index.js#L162

Next, you can do/run/start whatever you need in _startFn_:
image

And the last, there is no setup.js in quoting-service, the start process of API was initially implemented without it:
https://github.com/mojaloop/quoting-service/blob/main/src/server.js#L201

So i didn't change anything, and I tried to follow the same approach for starting handlers (using start fn). So could you, please, tell me what's wrong with this approach, and why do we need to change existing in this service approach?

File: src/handlers/QuotingHandler.js
Comment (geka-evk) [reply src/handlers/QuotingHandler.js:24]: Sure, let's add it.
But one thing I'd like to mention:
handling of any message is done by appropriate QuotesModel/BulkQuotesModel methods. And each such methods already has Metrics.getHistogram logic. So I just want to double check do we really need to add additional Metrics.getHistogram here? 
image

File: jest.config.js
Comment (vijayg10) [reply jest.config.js:11]: Yes, we need to add test coverage for all the code we added and we shouldn't exclude any file when we merge to master.

File: src/handlers/QuotingHandler.js
Comment (vijayg10) [reply src/handlers/QuotingHandler.js:24]: If there is no other functionality in the handler, then its fine. If there is any additional functionality in the handler on top of model, then we need to add metrics.

File: jest.config.js
Comment (geka-evk) [reply jest.config.js:11]: Ok, clear. Added a new task:

File: src/handlers/QuotingHandler.js
Comment (vijayg10) [reply src/handlers/QuotingHandler.js:55]: No, that's not what I am proposing. We can have separate function for each queue. I am just saying, instead of comparing the topic name, we can have event / action to determine the type of message like we are doing in central-ledger.

Comment (vijayg10) [reply src/handlers/QuotingHandler.js:55]: Oh wait, I think you are right. I don't see any reason to determine the message based on the topic name. I think its fine to have this logic here. But lets keep the event and action metadata in the kafka message just incase. You might already doing that, but just a reminder.

File: src/handlers/QuotingHandler.js
Comment (vijayg10) [reply src/handlers/QuotingHandler.js:83]: We need to reduce amount of if conditions for performance, I found more than one if condition in the flow to check if it is an error callback. That's why I proposed having a new function and calling a separate function based on the message reduces the need of further if conditions in the flow.

File: src/handlers/init.js
Comment (vijayg10) [reply src/handlers/init.js:33]: If its not there, lets add it to align it with central-ledger so that we can add if there are any initialisations later. Just like you added startFn, I am proposing to use the same naming convention and structure across different repos. If you think it takes time, then that's ok, we can standardise it later to align with central-ledger.

File: config/default.json
Comment (vijayg10) [reply config/default.json:155]: Oh ok. Then its up to you how you want to call it. I am fine with anything, because there is no standard.

File: src/api/bulkQuotes/{id}/error.js
Comment (vijayg10) [reply src/api/bulkQuotes/{id}/error.js:88]: Oh I got it why you removed from here.
But we need to audit in both places, here and in consumer.

File: jest.config.js
Comment (vijayg10) [reply jest.config.js:11]: Sure I am ok with the new task. You can add the coverage to the same PR.

File: src/handlers/QuotingHandler.js
Comment (geka-evk) [reply src/handlers/QuotingHandler.js:83]: If you remember, exactly for such reason (to avoid unnecessary checks) I proposed to have a separate queue for each payloads (even for _PUT success_ and _PUT error_). But we agreed to publish PUT payloads to one queue. In this case I don't see a good way to avoid 2 checks (first, check queue, then check a payload).

I think, we can avoid any condition check in our flow, if we'll specify for each topic a particular function (e.g., handlePostQuotes for topic-quotes-post and so on). In that case we won't need _handleMessages_ and _defineHandlerByTopic_ at all.
here's where we an do it: https://github.com/mojaloop/quoting-service/pull/321/files#diff-db88d0fcec394bf32e40eff7c738649415d22ee74c4c56577c8814c0c599ff6dR21

Lt's have a call and discuss if we need to do such optimization

File: src/handlers/init.js
Comment (geka-evk) [reply src/handlers/init.js:33]: I would suggest, first of all, to check GP tests for now, and make them pass. Coz it's the most important thing, as for me. After that we can have a separate call (with all devs), and discuss which exactly parts of our services we need to standardize

File: src/handlers/QuotingHandler.js
Comment (geka-evk) [reply src/handlers/QuotingHandler.js:24]: I've tried to make handlers as much simple as it's possible. Coz the main logic is inside models

image

File: src/api/bulkQuotes/{id}/error.js
Comment (geka-evk) [reply src/api/bulkQuotes/{id}/error.js:88]: @vijayg10 , do you mean, we need to have span.audit() for API handlers where we now just publish a message?
image

File: src/api/bulkQuotes.js
Comment (geka-evk) [reply src/api/bulkQuotes.js:63]: https://github.com/mojaloop/quoting-service/pull/321/files#diff-c1a25e85b3bda026d926f4885a558659679587b8e4014eeef189967bcefc809dR49

But it does `Logger.isDebugEnabled &&` check under the hood. Why do we need to repeat this logic all over the rest code?

File: src/api/bulkQuotes/{id}.js
Comment (geka-evk) [reply src/api/bulkQuotes/{id}.js:29]: If it's needed, I'll add it. But personally for me it's not so important to be listed as maintainers )

File: src/api/bulkQuotes/{id}/error.js
Comment (geka-evk) [reply src/api/bulkQuotes/{id}/error.js:75]: To be honest, I wasn't sure how it's better to handle such errors during publishing a message. But if you think we can use _model.handleException_ here, I'll update these parts of API handlers

File: src/api/bulkQuotes/{id}.js
Comment (geka-evk) [reply src/api/bulkQuotes/{id}.js:57]: But the actual processing will be done only when consumer picks up this message, and finish its processing. But this timer will be stopped just right after publishing.  Do you think, it still make sense to sure term "Process"?

File: package.json
Comment (kleyow) [inline package.json:59]: Any reason as to why `--runInBand` was removed?
Sometimes tests fail due to be ran in parallel, not saying it happens here but it's typically a preventive measure so we don't have stupid CI failures.

File: src/api/bulkQuotes/{id}.js
Comment (kleyow) [inline src/api/bulkQuotes/{id}.js:105]: Believe we would just rethrow the error and have a http sycronous 500 be sent back.
The kafka handler should handle the sending of the error callback notification if something goes wrong processing wise.

File: src/api/bulkQuotes/{id}/error.js
Comment (kleyow) [reply src/api/bulkQuotes/{id}/error.js:75]: Reposting here since I believe this is how the ml-api-adapter handles it.

Believe we would just rethrow the error and have a http syncronous 500 be sent back if something goes wrong publishing the message.
The kafka handler should handle the sending of the error callback notification if something goes wrong processing wise.

File: src/api/bulkQuotes.js
Comment (geka-evk) [reply src/api/bulkQuotes.js:103]: ok, agree. Will put it back.

File: package.json
Comment (geka-evk) [reply package.json:59]: the only reason to make them run faster.

Comment (kleyow) [reply package.json:59]: iirc mostly random. Yes you're abosulutely correct that it's the test themselves are not properly isolated/reset.

If you don't run into any problems I'm fine with removing it.

File: src/api/bulkQuotes/{id}.js
Comment (geka-evk) [reply src/api/bulkQuotes/{id}.js:105]: ...rethrow the error and have a http syncronous 500 be sent back...
@kleyow, Initially I also was thinking about it, but when I checked Swagger API docs, I didn't find 500 status code among allowed response for quotes/bulkQuotes (only 501 and 503, which is not really suitable for publishing error, IMHO).

File: src/api/bulkQuotes/{id}/error.js
Comment (geka-evk) [reply src/api/bulkQuotes/{id}/error.js:75]: @vijayg10 , what do you think regarding Kevin's suggestion?  Let's make a final decision in which way I should handle publishing error

File: src/api/bulkQuotes.js
Comment (geka-evk) [reply src/api/bulkQuotes.js:88]: the same reason as here: https://github.com/mojaloop/quoting-service/pull/321#discussion_r1425125175

@vijayg10 , a question: https://github.com/mojaloop/quoting-service/pull/321#discussion_r1425334806

File: src/api/bulkQuotes.js
Comment (geka-evk) [reply src/api/bulkQuotes.js:68]: as checked it, and I see that it uses `getFunctionalityAction()` and topic templates for making topicName. I'm not sure we need it in this case, when we have particular topic names

File: package.json
Comment (geka-evk) [reply package.json:90]: sure, will do

File: src/api/bulkQuotes/{id}/error.js
Comment (vijayg10) [reply src/api/bulkQuotes/{id}/error.js:88]: Yes that's right. We need audit before publishing a messages in api service and one more audit after consuming that message in handler service.

File: src/handlers/QuotingHandler.js
Comment (vijayg10) [reply src/handlers/QuotingHandler.js:24]: Ok, then lets leave metrics for now

Comment (vijayg10) [reply src/handlers/QuotingHandler.js:24]: @kleyow what's your suggestion on this with your performance hat on?

File: src/api/bulkQuotes/{id}.js
Comment (kleyow) [reply src/api/bulkQuotes/{id}.js:105]: Hmm...I'm just only now realizing that. I'm not sure how to answer now.

File: src/handlers/QuotingHandler.js
Comment (kleyow) [reply src/handlers/QuotingHandler.js:24]: To me more metrics is better, there have been times I realize we don't have granular enough metrics and having to add them in is a pain. 

But I can understand the flipside that the handler is so bare that metrics aren't needed. I'm fine with either.

**Comment by kleyow** [comment]
Lgtm. Just the CI, dep/audit updating discussed on slack.

**Review by vijayg10 — APPROVED** [review]

**Pull request merged**
